File: Top 100 Codes/03-Codes for Recursion/19-Given a string s, remove all its adjacent duplicate characters recursively/Givenastringsremoveallitsadjacentduplicatecharactersrecursively.py
def remove_adjacent_duplicates(s):
    n = len(s)
    if n < 2:
        return s

    result = []
    i = 0

    while i < n:
        # Check for adjacent duplicates
        if i < n - 1 and s[i] == s[i + 1]:
            # Skip all adjacent duplicates
            while i < n - 1 and s[i] == s[i + 1]:
                i += 1
        else:
            result.append(s[i])
        i += 1

    # Convert list back to string
    new_string = ''.join(result)

    # If no change, return the string
    if new_string == s:
        return new_string

    # Recursively call for further reduction
    return remove_adjacent_duplicates(new_string)


# Test the function
s = input("Enter a string: ")
result = remove_adjacent_duplicates(s)
print("Result after removing adjacent duplicates:", result)

File: Top 100 Codes/03-Codes for Recursion/19-Given a string s, remove all its adjacent duplicate characters recursively/Givenastringsremoveallitsadjacentduplicatecharactersrecursively.cpp
#include <iostream>
#include <string>
using namespace std;

// Function to recursively remove adjacent duplicates
string removeAdjacentDuplicates(string s) {
    int n = s.length();
    if (n < 2) return s;

    string result = "";
    int i = 0;

    while (i < n) {
        // Check for adjacent duplicates
        if (i < n - 1 && s[i] == s[i + 1]) {
            // Skip all adjacent duplicates
            while (i < n - 1 && s[i] == s[i + 1]) {
                i++;
            }
        } else {
            // Add non-duplicate character to result
            result += s[i];
        }
        i++;
    }

    // If no change is made, return the result
    if (result.length() == s.length()) {
        return result;
    }

    // Recursively call for further reduction
    return removeAdjacentDuplicates(result);
}

int main() {
    string s;
    cout << "Enter a string: ";
    cin >> s;
    string result = removeAdjacentDuplicates(s);
    cout << "Result after removing adjacent duplicates: " << result << endl;
    return 0;
}
